spec/abci/abci++_basic_concepts.md

```diff
@@ -206,19 +206,27 @@ More details on managing state across connections can be found in the section on
 
 ## Proposal related timeouts
 
-`PrepareProposal` and `ProcessProposal` stand on the consensus algorithm critical path,
-i.e., CometBFT cannot make progress while these methods are being executed.
-Hence, if the Application takes a long time preparing a proposal or in verifying it,
-the default values of *TimeoutPropose* and *TimeoutPrevote* might not be sufficient
-to accommodate these methods' executions and validator nodes might time out and prevote `nil`.
+`PrepareProposal` stands on the consensus algorithm critical path,
+i.e., CometBFT cannot make progress while this method is being executed.
+Hence, if the Application takes a long time preparing a proposal,
+the default value of *TimeoutPropose* might not be sufficient
+to accommodate the method's execution and validator nodes might time out and prevote `nil`.
 The proposal, in this case, will probably be rejected and a new round will be necessary.
 If this happens successively, the liveness of CometBFT might be compromised.
 
-This is particularly important if applications need to execute the block to validate it.
-Then they would have to execute it in `PrepareProposal` and `ProcessProposal`.
 
-Operators will need to adjust the default values of *TimeoutPropose* and *TimeoutPrevote* in CometBFT's
-configuration file, in order to suit the needs of the particular application being deployed.
+Operators will need to adjust the default value of *TimeoutPropose* in CometBFT's configuration file,
+in order to suit the needs of the particular application being deployed.
+
+This is particularly important if applications implement *immediate execution*, a technique in
+which applications require the proposer to execute a block in `PrepareProposal` (and all validators to execute it in `ProcessProposal`) to ensure that all transactions are executable in case the block is decided.
```

Comment:
Can you break this line please?

Reply:
Actually, the validators have to verify the block execution in `ProcessProposal`.

Reply:
I didn't understand the second part of the sentence, starting from "to ensure that..."

Reply:
We need to agree on how to break lines, if at the 80 or 100 char mark or something else.
I prefer breaking into periods, even if it makes lines longer.

I didn't catch this comment.
> Actually, the validators have to verify the block execution in ProcessProposal.

Do you mean that I should say validate instead of execute?
I guess I just expanded "validate" to "execute a block (...) to ensure all transactions are executable in case the block is decided."

Reply:
We can have two cases here:
1. Everyone executes the block in `ProcessProposal`, as a pre-execution that will actually be confirmed when the block is finally committed
2. The proposes execute the block on `PrepareProposal`, encodes somehow the results of the execution on the block itself, and non-proposers have to check whether the proposer's execution is valid, in `ProcessProposal`

I think we should discourage option 2, because it adds a lot of complexity and corner cases. My suggestion can be just to mention case 1 here.

Reply:
We need to agree on how to break lines, if at the 80 or 100 char mark or something else.
I prefer breaking into periods, even if it makes lines longer.

I didn't catch this comment.
> Actually, the validators have to verify the block execution in ProcessProposal.

Do you mean that I should say validate instead of execute?
I guess I just expanded "validate" to "execute a block (...) to ensure all transactions are executable in case the block is decided."

Reply:
> to ensure that all transactions are executable in case the block is decided.

This part I really don't get: are we executing txs to ensure they can be executed? If we are executing them now, we don't execute them again upon Commit. Upon commit we only check whether the block that we previously executed is the same block that has been committed. If this is not the case, we execute the block on Commit replacing the (optimistic) execution we performed upon `ProcessProposal`.